### ISSUE TITLE
New "Precompiled classes" configurable option

### ISSUE THREAD

**Issue by acslf09** [hclproducts/appscan-source-gradle-plugin#20]
Requesting to add "Precompiled classes" configurable option which will set "Properties -> Project Dependencies -> Optimizations -> Precompiled classes" option to true and points to compiled files location.

**Comment by mattmurp** [comment]
Thanks for the feedback.  By default the plugin will scan the .class files compiled by Gradle by setting the source root to the location of the .class files.  If that doesn't satisfy your needs, can you please explain why?

**Comment by acslf09** [comment]
Thanks for your prompt response. We've looked into several options (mentioned below) with HCL support team and reached to a conclusion that adding "Precompiled classes" location provide comprehensive scan with least manual steps.

1. By default, AppScan is performing "Pattern Analysis" on class files and not performing "Framework Analysis" on either class or java source files. Therefore, not providing comprehensive scan result.

2. On setting "class" extension to scan for each project, AppScan does perform both Framework and Pattern analysis but unable to link .class & .java files automatically. We've to manually find & select each source file where the issues are detected.

3. We've also tried building code using "compilecode = true" option. Because applications uses complex build structure, AppScan is unable to compile it and throwing compilation error.

The requested option will help automate the process for several applications containing multiple projects.

**Comment by acslf09** [comment]
Any updates?

**Comment by mattmurp** [comment]
Thanks for the explanation.  So the main benefit of this enhancement is linking the .class files to the .java files when viewing the assessment.  AppScan Source should help with this, in that once you've searched for a .java file it will remember that location and try to locate other files using it.  While not perfect, that should limit the number of times a user needs to browse to locate source files.
We'll have to think about how we could implement this enhancement.  The challenge is that with the "Precompiled Classes" option there can be only 1 location containing the .class files.  For projects with multiple sourcesets that have different output locations, this won't work.  We would need to create a project per sourceset in such cases.

**Comment by mattmurp** [comment]
Unfortunately, using the parent location in the case of multiple source sets will not work with the Precompiled Classes option.  The location specified for that option must match the package hierarchy of the source files that were specified.
To be clear, even without this enhancement the plugin does not require recompiling the application (though there is an option to do so), so the only gain is the automatic mapping to the source files. As mentioned above, once you've browsed to 1 source file, AppScan Source uses that location to locate other sources.  Additionally, if the plugin is being used to run a scan and the assessment is viewed on other systems, the user will still have to browse to the first source file (unless the absolute paths to all files are identical on both systems).
We'll have to investigate how this enhancement could be implemented or if there is a better way to achieve the automatic mapping.

**Comment by acslf09** [comment]
Just to confirm, when you say "plugin does not require recompiling the application" are you referring to an option to scan "class" extension for each project?
Because that also involve manual work to add this extension for multiple projects. This is can done globally by editing "file_extensions.xml" but that'll affect other scans running simultaneously on that machine which does require recompilation.

If automatic mapping and automatic option to scan "class" extension (without setting globally) can be achieved then we might not need this enhancement.

**Comment by mattmurp** [comment]
Yes, the default behavior of the plugin is to scan the .class files that were generated by the Gradle build.  And yes, you do need to set the .class extension to be scanned as a prerequisite.  It's recommended to do so at the global level.  Doing so simply sets Java projects to scan any .class files that are in the given source root(s).  This does not in any way effect projects that do not contain .class files in the source root(s).

**Comment by acslf09** [comment]
My concern with that is why class extension scan is not set _by default_ if the project does/not contain class files.
